hal/halx86/apic/apic.c

```diff
@@ -671,87 +858,470 @@ HalDisableSystemInterrupt(
+{
+    PHALP_PCR_HAL_RESERVED HalReserved;
+    ULONG EFlags;
+
+    EFlags = __readeflags();
+    KeGetPcr()->Irql = Irql;
+
+    if (!(EFlags & EFLAGS_INTERRUPT_MASK))
+        _enable();
+
+    HalReserved = (PHALP_PCR_HAL_RESERVED)KeGetPcr()->HalReserved;
+    if (Irql == APC_LEVEL)
+    {
+        HalReserved->ApcRequested = FALSE;
+        KiDeliverApc(0, 0, (PKTRAP_FRAME)TrapFrame);
+    }
+    else if (Irql == DISPATCH_LEVEL)
+    {
+        HalReserved->DpcRequested = FALSE;
+        KiDispatchInterrupt();
+    }
+    else
+    {
+        DbgBreakPoint();
+    }
+
+    if (!(EFlags & EFLAGS_INTERRUPT_MASK))
+        _disable();
+}
+
+VOID
+FASTCALL
+HalpCheckForSoftwareInterrupt(_In_ KIRQL NewIrql,
+                              _In_ PKTRAP_FRAME TrapFrame)
+{
+    PHALP_PCR_HAL_RESERVED HalReserved;
+    BOOLEAN ApcRequested;
+    BOOLEAN DpcRequested;
+
+    HalReserved = (PHALP_PCR_HAL_RESERVED)KeGetPcr()->HalReserved;
+
+    ApcRequested = HalReserved->ApcRequested;
+    DpcRequested = HalReserved->DpcRequested;
+
+    if (NewIrql)
+    {
+        if (NewIrql == APC_LEVEL && HalReserved->DpcRequested)
+        {
+            do
+            {
+                HalpDispatchSoftwareInterrupt(DISPATCH_LEVEL, TrapFrame);
+                KeSetCurrentIrql(APC_LEVEL);
+                HalReserved = (PHALP_PCR_HAL_RESERVED)KeGetPcr()->HalReserved;
+            }
+            while (HalReserved->DpcRequested);
+        }
+
+        return;
+    }
+
+    while (ApcRequested | DpcRequested)
+    {
+        if (DpcRequested)
+        {
+            HalpDispatchSoftwareInterrupt(DISPATCH_LEVEL, TrapFrame);
+        }
+        else
+        {
+            HalpDispatchSoftwareInterrupt(APC_LEVEL, TrapFrame);
+        }
+
+        KeSetCurrentIrql(PASSIVE_LEVEL);
+
+        HalReserved = (PHALP_PCR_HAL_RESERVED)KeGetPcr()->HalReserved;
+        ApcRequested = HalReserved->ApcRequested;
+        DpcRequested = HalReserved->DpcRequested;
+    }
+}
+
+BOOLEAN
+NTAPI
+HalBeginSystemInterrupt(_In_ KIRQL NewIrql,
+                        _In_ ULONG SystemVector,
+                        _Out_ PKIRQL OutOldIrql)
+{
+    PUCHAR pPrcbVector;
+    UCHAR Idx;
+    KIRQL OldIrql;
+
+    OldIrql = KeGetCurrentIrql();
+
+    if (OldIrql < HalpVectorToIRQL[(UCHAR)SystemVector >> 4])
+    {
+        *OutOldIrql = OldIrql;
+        KeSetCurrentIrql(NewIrql);
+
+        _enable();
+        return TRUE;
+    }
+
+    pPrcbVector = (PUCHAR)KeGetCurrentPrcb()->HalReserved;
+
+    Idx = pPrcbVector[0];
+    pPrcbVector[Idx + 1] = SystemVector;
+    pPrcbVector[0] = Idx + 1;
+
+    _enable();
+    return FALSE;
+}
+
+static
+VOID
+NTAPI
+HalpEndSystemInterrupt(_In_ PHAL_INTERRUPT_CONTEXT IntContext,
+                       _In_ KIRQL OldIrql)
+{
+    KIRQL Irql = HalpVectorToIRQL[IntContext->Vector >> 4];
+
+    if (Irql < KeGetCurrentIrql())
+    {
+        HalpLowerIrqlHardwareInterrupts(Irql);
+    }
+
+    if (ApicRead(APIC_PPR) != (IntContext->Vector & 0xF0))
```

Comment:
From the Intel SDM (Vol. 3A, 10-29)
![Screenshot_20210412_022213](https://user-images.githubusercontent.com/961957/114324953-f15e6780-9b35-11eb-8696-bc977a3d3db0.png)


So you're touching an undocumented area here. What do you expect to see? Does it work on all platforms?

Reply:
Fixed

Reply:
What was changed?
`(ApicRead(APIC_PPR) & 0xF0) != (IntContext->Vector & 0xF0)` still reads reserved area in PPR

Reply:
1. The reserved area is no longer used for comparison.
2. Does it say somewhere that this area cannot be read?